drivers: watchdog: litex: add litex watchdog

maass-hamburg opened the pull request:
this adds a driver for the litex watchdog.

on LiteX: https://github.com/enjoy-digital/litex/pull/1996

maass-hamburg committed:
soc: riscv: litex: soc.h: add litex_write64

add function for a 64-bit write in litex.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

maass-hamburg committed:
drivers: watchdog: litex: add litex watchdog

this adds a driver for the litex watchdog.

Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>

fkokosinski commented:
Hey, thanks for this PR. In general the code LGTM.

That being said, the watchdog driver is not being used by the `litex_vexriscv` platform (or any other in-tree platform), meaning it will not be built/tested by twister in the CI. This essentially means that from the CI standpoint this is dead code.

Could you perhaps add it to the `litex_vexriscv` platform and do the RTL configuration repo that this platform is targeting? Thanks

fkokosinski commented on drivers/interrupt_controller/intc_vexriscv_litex.c:
```diff
@@ -58,6 +58,12 @@ static inline void vexriscv_litex_irq_setie(uint32_t ie)
 	}
 }
 
+#define IRQ_FKT(n)                                                                                 \
```
The name of this macro is somewhat ambiguous

fkokosinski commented on drivers/interrupt_controller/intc_vexriscv_litex.c:
```diff
@@ -103,6 +109,10 @@ static void vexriscv_litex_irq_handler(const void *device)
 	}
+
+#ifdef CONFIG_WDT_LITEX
+	DT_FOREACH_STATUS_OKAY(litex_watchdog, IRQ_FKT);
```
Could you apply a similar change to other IRQs handled in this routine?

maass-hamburg replied:
changed it to use `DT_FOREACH_STATUS_OKAY_NODE`. That way the interrupt controller doesn't has to be modified anymore, when new peripherals are added. This way we can also support custom out off tree peripherals and drivers with interrupts without having to edit this.

maass-hamburg commented:
> RTL configuration repo

It is now added to the `litex_vexriscv` platform. I also added the watchdog here: https://github.com/litex-hub/zephyr-on-litex-vexriscv/pull/21

ycsin commented on drivers/watchdog/wdt_litex.c:
```diff
@@ -0,0 +1,238 @@
+	.setup = wdt_litex_setup,
+	.disable = wdt_litex_disable,
+	.install_timeout = wdt_litex_install_timeout,
+	.feed = wdt_litex_feed
```
```suggestion
	.feed = wdt_litex_feed,
```

maass-hamburg commented on drivers/interrupt_controller/intc_vexriscv_litex.c:
```diff
@@ -58,6 +48,14 @@ static inline void vexriscv_litex_irq_setie(uint32_t ie)
+		ite->isr(ite->arg);                                                                \
+	}
+
+#define LITEX_IRQ_ADD(n) IF_ENABLED(DT_IRQ_HAS_IDX(n, 0), (LITEX_IRQ_ADD_HELPER(n)))
```
Litex only has one interrupt controller and this interrupt controller driver also only supports one instance. There shouldn't be any other interrupt controller.

fkokosinski commented on drivers/watchdog/wdt_litex.c:
```diff
@@ -0,0 +1,238 @@
+/*
+ * Copyright (C) 2017 Intel Corporation
```
The copyright seems to be wrong

maass-hamburg replied:
Good catch ;)

maass-hamburg commented:
ping

maass-hamburg commented:
ping @tgorochowik @kgugala

maass-hamburg commented:
litte reminder for the second review @tgorochowik @kgugala

carlescufi commented:
@maass-hamburg please rebase

maass-hamburg commented:
rebase done to resolve conflicts

maass-hamburg commented:
a `};` got lost in the dtsi during rebase

maass-hamburg commented:
@fkokosinski @tgorochowik litte reminder to re-approve after the rebase

tgorochowik commented on drivers/watchdog/wdt_litex.c:
```diff
@@ -0,0 +1,238 @@
+	struct wdt_litex_data *data = dev->data;
+	uint32_t control;
+
+	if (config->cycles_size <= 4) {
+		litex_write32(k_ms_to_cyc_floor32(data->timeout), config->cycles_addr);
+	} else {
+		litex_write64(k_ms_to_cyc_floor64(data->timeout), config->cycles_addr);
+	}
```
was this tested? You seem to be using bits as size in the corresponding litex PR: https://github.com/enjoy-digital/litex/pull/1996/files#diff-91072e8d1728a622115f31f10d4a9a6902f27f22054a6236b4b09e00aa97287aR23

Or is it converted to bytes in litex somewhere else

maass-hamburg replied:
It is the size of the reg in the dts, which is in bytes

tgorochowik replied:
oh, right

aescolar commented:
@maass-hamburg  rebase needed

maass-hamburg commented:
Rebased to resolve conflicts
